### PR TITLE
TorchHubRegistery

### PR DESCRIPTION
**Pull request by da-tubi**
``` sql
CREATE MODEL yolov5
FLAVOR yolov5
PREPROCESSOR 'rikai.contrib.yolov5.transforms.pre_processing'
POSTPROCESSOR 'rikai.contrib.yolov5.transforms.post_processing'
OPTIONS (device="cpu", batch_size=32)
RETURNS {OUTPUT_SCHEMA}
USING "torchhub:///ultralytics/yolov5:v6.0/yolov5s";

select ML_PREDICT(yolov5, "path/to/lena.png")
```

Currently, only yolov5's pre/post processors are available.

### CODE REVIEW

File: src/main/scala/ai/eto/rikai/sql/model/Registry.scala [hunk 1]
```diff
@@ -57,7 +79,8 @@ private[rikai] object Registry {
     */
   val DEFAULT_REGISTRIES = Map(
     "rikai.sql.ml.registry.file.impl" -> "ai.eto.rikai.sql.model.fs.FileSystemRegistry",
-    "rikai.sql.ml.registry.mlflow.impl" -> "ai.eto.rikai.sql.model.mlflow.MlflowRegistry"
+    "rikai.sql.ml.registry.mlflow.impl" -> "ai.eto.rikai.sql.model.mlflow.MlflowRegistry",
```
Comment (eddyxu) [inline src/main/scala/ai/eto/rikai/sql/model/Registry.scala:82]: This is amazing.  How much work is there to make Rikai to support multiple model registries?

File: src/main/scala/ai/eto/rikai/sql/model/torchhub/TorchHubRegistry.scala [hunk 2]
```diff
@@ -0,0 +1,11 @@
+package ai.eto.rikai.sql.model.torchhub
```
Comment (eddyxu) [inline src/main/scala/ai/eto/rikai/sql/model/torchhub/TorchHubRegistry.scala:1]: Lets add license here?

File: python/tests/spark/sql/codegen/test_torchhub_registry.py [hunk 3]
```diff
@@ -0,0 +1,21 @@
+from pyspark.sql import SparkSession
```
Comment (eddyxu) [inline python/tests/spark/sql/codegen/test_torchhub_registry.py:1]: lets add license here?

File: src/main/scala/ai/eto/rikai/sql/model/Registry.scala (hunk 1 above)
Comment (da-tubi) [reply src/main/scala/ai/eto/rikai/sql/model/Registry.scala:82]: I took about 1d to create this pull request for the torchhub registry.

Comment (darcy-shen) [reply src/main/scala/ai/eto/rikai/sql/model/Registry.scala:82]: For multiple model registries, I created a issue: https://github.com/eto-ai/rikai/issues/377

We need to design syntax first.

Comment (da-tubi) [reply src/main/scala/ai/eto/rikai/sql/model/Registry.scala:82]: For multiple model registries, I created a issue: #377

We need to design syntax first.

File: contrib/yolov5/tests/test_torchhub.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    image_path = f"{work_dir}/python/tests/assets/test_image.jpg"
+    spark.sql(
+        f"""
+CREATE MODEL yolov5
+FLAVOR yolov5
+PREPROCESSOR 'rikai.contrib.yolov5.transforms.pre_processing'
+POSTPROCESSOR 'rikai.contrib.yolov5.transforms.post_processing'
+OPTIONS (device="cpu", batch_size=32)
+RETURNS {OUTPUT_SCHEMA}
+USING "torchhub:///ultralytics/yolov5:v6.0/yolov5s";
```
Comment (da-tubi) [inline contrib/yolov5/tests/test_torchhub.py:33]: This why I'm using `yolov5:v6.0`, seems to be a yolov5 bug.
```
(rikai) ➜  python git:(feature/torchhub) python
Python 3.8.10 (default, May 19 2021, 11:01:55)
[Clang 10.0.0 ] :: Anaconda, Inc. on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import torch
>>> torch.hub.load("ultralytics/yolov5:v5.0", "yolov5s")
Using cache found in /Users/da/.cache/torch/hub/ultralytics_yolov5_v5.0

                 from  n    params  module                                  arguments
  0                -1  1      3520  models.common.Focus                     [3, 32, 3]
  1                -1  1     18560  models.common.Conv                      [32, 64, 3, 2]
  2                -1  1     18816  models.common.C3                        [64, 64, 1]
  3                -1  1     73984  models.common.Conv                      [64, 128, 3, 2]
  4                -1  1    156928  models.common.C3                        [128, 128, 3]
  5                -1  1    295424  models.common.Conv                      [128, 256, 3, 2]
  6                -1  1    625152  models.common.C3                        [256, 256, 3]
  7                -1  1   1180672  models.common.Conv                      [256, 512, 3, 2]
  8                -1  1    656896  models.common.SPP                       [512, 512, [5, 9, 13]]
  9                -1  1   1182720  models.common.C3                        [512, 512, 1, False]
 10                -1  1    131584  models.common.Conv                      [512, 256, 1, 1]
 11                -1  1         0  torch.nn.modules.upsampling.Upsample    [None, 2, 'nearest']
 12           [-1, 6]  1         0  models.common.Concat                    [1]
 13                -1  1    361984  models.common.C3                        [512, 256, 1, False]
 14                -1  1     33024  models.common.Conv                      [256, 128, 1, 1]
 15                -1  1         0  torch.nn.modules.upsampling.Upsample    [None, 2, 'nearest']
 16           [-1, 4]  1         0  models.common.Concat                    [1]
 17                -1  1     90880  models.common.C3                        [256, 128, 1, False]
 18                -1  1    147712  models.common.Conv                      [128, 128, 3, 2]
 19          [-1, 14]  1         0  models.common.Concat                    [1]
 20                -1  1    296448  models.common.C3                        [256, 256, 1, False]
 21                -1  1    590336  models.common.Conv                      [256, 256, 3, 2]
 22          [-1, 10]  1         0  models.common.Concat                    [1]
 23                -1  1   1182720  models.common.C3                        [512, 512, 1, False]
 24      [17, 20, 23]  1    229245  models.yolo.Detect                      [80, [[10, 13, 16, 30, 33, 23], [30, 61, 62, 45, 59, 119], [116, 90, 156, 198, 373, 326]], [128, 256, 512]]
Model Summary: 283 layers, 7276605 parameters, 7276605 gradients, 17.1 GFLOPS

Downloading https://github.com/ultralytics/yolov5/releases/download/v6.0/yolov5s.pt to yolov5s.pt...
```

Comment (da-tubi) [reply contrib/yolov5/tests/test_torchhub.py:33]: v5.0 code are used, but v6.0 yolov5s.pt are downloaded.

File: python/rikai/spark/sql/codegen/torchhub_registry.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+            },
+        }
+
+        self.repo_or_dir = repo_or_dir
+        self.model = model
+        super().__init__(spec=spec, validate=True)
+
+    def load_model(self) -> Any:
+        """Load the model artifact specified in this spec"""
+        return torch.hub.load(self.repo_or_dir, self.model)
```
Comment (da-tubi) [inline python/rikai/spark/sql/codegen/torchhub_registry.py:46]: useful options can be cutomized here like:

``` python
torch.hub.load(self.repo_or_dir, self.model, reload=True)
```
This can be done be the follow up pull requests.

File: python/tests/spark/sql/codegen/test_torchhub_registry.py [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+from rikai.contrib.torch.transforms.fasterrcnn_resnet50_fpn import (
+    OUTPUT_SCHEMA,
+)
+
+
+def test_create_model(rikai_spark: SparkSession):
+    # TODO: run ml_predict on resnet50
+    # rikai.contrib.torch.transforms.fasterrcnn_resnet50_fpn
+    # does not actually work for torchhub loaded model
```
Comment (da-tubi) [inline python/tests/spark/sql/codegen/test_torchhub_registry.py:25]: where to put torchhub transforms pre/post precessors, this needs to be dicuessed.

File: python/tests/spark/sql/codegen/test_torchhub_registry.py [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+from rikai.contrib.torch.transforms.fasterrcnn_resnet50_fpn import (
+    OUTPUT_SCHEMA,
+)
+
+
+def test_create_model(rikai_spark: SparkSession):
+    # TODO: run ml_predict on resnet50
+    # rikai.contrib.torch.transforms.fasterrcnn_resnet50_fpn
+    # does not work for torchhub loaded model
```
Comment (da-tubi) [inline python/tests/spark/sql/codegen/test_torchhub_registry.py:25]:
```
    resnet = torchvision.models.detection.fasterrcnn_resnet50_fpn(
        pretrained=True,
        progress=False,
    )

resnet2 = torch.hub.load("pytorch/vision:v0.9.1", "resnet50")
```
resnet and resnet2 are not the same. ML_PREDICT on torchhub loaded model are tested on the `rikai-yolov5` contrib.

File: python/tests/conftest.py [hunk 8]
```diff
@@ -229,6 +229,31 @@ def s3_tmpdir() -> str:
         logging.warn("Could not delete directory: %s", s3fs_path)
 
 
+@pytest.fixture(scope="session")
```
Comment (darcy-shen) [inline python/tests/conftest.py:232]: too slow to use the default test SparkSession, we need to do some cleanup in another pull request

Comment (da-tubi) [inline python/tests/conftest.py:232]: too slow to use the default test SparkSession, we need to do some cleanup in another pull request

Comment (eddyxu) [reply python/tests/conftest.py:232]: Could you clarify why it is slow to use SparkSession? where is the part which takes the most of the time?

File: python/rikai/spark/sql/codegen/torchhub_registry.py [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+        uri = raw_spec.getUri()
+        logger.info(f"Resolving model {name} from {uri}")
+        parsed = urlparse(uri)
+        if parsed.netloc:
+            raise ValueError(
+                "URI with 2 forward slashes is not supported, "
+                "try URI with 1 slash instead"
+            )
+        if not parsed.scheme:
+            raise ValueError("Scheme must be torchhub. How did you get here?")
```
Comment (eddyxu) [inline python/rikai/spark/sql/codegen/torchhub_registry.py:70]: Lets rephrase this error to "Expect schema: torchhub, but got {parsed.schema}"? Lets make the warning message less aggresive.

File: python/rikai/spark/sql/codegen/torchhub_registry.py [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+        parts = parsed.path.strip("/").split("/")
+        repo_or_dir = "/".join(parts[0:-1])
```
Comment (eddyxu) [inline python/rikai/spark/sql/codegen/torchhub_registry.py:72]: Can just be `parts[:-1]`

Comment (eddyxu) [reply python/rikai/spark/sql/codegen/torchhub_registry.py:72]: Could you add a few tests for bad forms of url?

File: contrib/yolov5/tests/test_torchhub.py [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+from rikai.contrib.yolov5.transforms import OUTPUT_SCHEMA
+
+
+def test_yolov5(spark: SparkSession):
+    work_dir = Path().absolute().parent.parent
+    image_path = f"{work_dir}/python/tests/assets/test_image.jpg"
+    spark.sql(
+        f"""
+CREATE MODEL yolov5
+FLAVOR yolov5
```
Comment (eddyxu) [inline contrib/yolov5/tests/test_torchhub.py:28]: Sorry I just catch this late in the game. So flavor here is `yolov5`, it means a model-special runner?

cc @changhiskhan is this desired behavior?

Comment (da-tubi) [reply contrib/yolov5/tests/test_torchhub.py:28]: See https://github.com/eto-ai/rikai/pull/376 and https://github.com/eto-ai/rikai/blob/master/contrib/yolov5/REAMDE.md

File: python/tests/conftest.py (hunk 8 above)
Comment (da-tubi) [reply python/tests/conftest.py:232]: `rikai_spark()` is the default SparkSession before https://github.com/eto-ai/rikai/pull/369

I could revert this part and just use `spark()` since I have tested it locally.

`spark()` is slow because of extra dependencies and mlflow initialization. It takes minutes for me to download jars from maven.

File: contrib/yolov5/tests/test_torchhub.py (hunk 11 above)
Comment (changhiskhan) [reply contrib/yolov5/tests/test_torchhub.py:28]: yeah, yolov5 has a bit of a weird interface so we're using a custom flavor here. @da-tubi a lot of people will look to the tests for examples so maybe let's add another one using the standard pytorch flavor? And in this test just comment why we need the custom flavor.

**Comment by changhiskhan** [comment]
@da-tubi this is very exciting. To improve the user-experience, is there a default way to set pre/post processing and the output schema?

**Comment by da-tubi** [comment]
> To improve the user-experience, is there a default way to set pre/post processing and the output schema?

Yes for pre/post processing, I will set the default one when working on https://github.com/eto-ai/rikai/issues/356

File: contrib/yolov5/tests/test_torchhub.py (hunk 4 above)
Comment (da-tubi) [reply contrib/yolov5/tests/test_torchhub.py:33]: That's why v5.0 is always downloading the latest *pt https://github.com/ultralytics/yolov5/blob/f5b8f7d54c9fa69210da0177fec7ac2d9e4a627c/utils/google_utils.py#L25

File: python/rikai/spark/sql/codegen/base.py [hunk 12]
```diff
@@ -209,3 +209,26 @@ def register_udf(spark: SparkSession, udf: Callable, name: str) -> str:
     spark.udf.register(func_name, udf)
     logger.info(f"Created model inference pandas_udf with name {func_name}")
     return func_name
+
+
+def codegen_from_spec(
+    spark: SparkSession, spec: dict, name: Optional[str] = None
+) -> str:
+    """Generate code from an MLFlow runid
```
Comment (changhiskhan) [inline python/rikai/spark/sql/codegen/base.py:217]: is this still specific to an mlflow model? does the torchhub model use this?

**Review by changhiskhan — APPROVED** [review]
lgtm just 1 minor docstring nit
